File: src/lib/__tests__/rankProjection.test.ts
import { projectFinalRank, gradeForRating, isOnTrackForGrade } from "../rankProjection"
import type { GradeThreshold, RankObservation } from "../rankProjection"

/** A small illustrative rank table (the real caller supplies the game's full RANK_MINS/RANK_LABELS). */
const grades: GradeThreshold[] = [
    { grade: "E", minRating: 2000 },
    { grade: "D", minRating: 3000 },
    { grade: "C", minRating: 4000 },
    { grade: "B", minRating: 5500 },
    { grade: "A", minRating: 7000 },
    { grade: "S", minRating: 9000 },
]

describe("gradeForRating", () => {
    test("returns the highest grade whose minimum is met", () => {
        expect(gradeForRating(4500, grades)).toBe("C")
        expect(gradeForRating(7000, grades)).toBe("A")
        expect(gradeForRating(9999, grades)).toBe("S")
    })

    test("returns the lowest band below every threshold", () => {
        expect(gradeForRating(0, grades)).toBe("E")
    })
})

describe("projectFinalRank", () => {
    test("extrapolates a steady climb to the final turn and up a grade", () => {
        // ~150 rating/turn; from 3831 at turn 55 to turn 73 -> +18*150 ≈ +2700 -> ~6500 -> grade B.
        const obs: RankObservation[] = [
            { turn: 45, rating: 2300 },
            { turn: 50, rating: 3050 },
            { turn: 55, rating: 3831 },
        ]
        const proj = projectFinalRank(obs, 55, 73, grades)!
        expect(proj.currentGrade).toBe("D")
        expect(proj.ratingPerTurn).toBeGreaterThan(140)
        expect(proj.projectedRating).toBeGreaterThan(6000)
        expect(proj.projectedGrade).toBe("B")
    })

    test("never projects below the current rating", () => {
        // A declining/noisy slope must not drag the projection under what's already banked.
        const obs: RankObservation[] = [
            { turn: 40, rating: 5000 },
            { turn: 50, rating: 5200 },
            { turn: 60, rating: 5100 },
        ]
        const proj = projectFinalRank(obs, 60, 73, grades)!
        expect(proj.projectedRating).toBeGreaterThanOrEqual(proj.currentRating)
    })

    test("a single observation yields a flat projection (no slope)", () => {
        const proj = projectFinalRank([{ turn: 60, rating: 5000 }], 60, 73, grades)!
        expect(proj.ratingPerTurn).toBe(0)
        expect(proj.projectedRating).toBe(5000)
        expect(proj.projectedGrade).toBe("C")
    })

    test("uses the latest observation at or before the current turn as current", () => {
        const obs: RankObservation[] = [
            { turn: 50, rating: 4000 },
            { turn: 55, rating: 4500 },
            { turn: 60, rating: 5000 },
        ]
        const proj = projectFinalRank(obs, 55, 73, grades)!
        expect(proj.currentRating).toBe(4500)
    })

    test("returns null with no observations", () => {
        expect(projectFinalRank([], 60, 73, grades)).toBeNull()
    })

    test("already at the final turn projects the current rating unchanged", () => {
        const obs: RankObservation[] = [
            { turn: 71, rating: 6800 },
            { turn: 73, rating: 7100 },
        ]
        const proj = projectFinalRank(obs, 73, 73, grades)!
        expect(proj.projectedRating).toBe(7100)
        expect(proj.projectedGrade).toBe("A")
    })
})

describe("isOnTrackForGrade", () => {
    test("true when the projection clears the target grade minimum", () => {
        const proj = projectFinalRank(
            [
                { turn: 45, rating: 2300 },
                { turn: 55, rating: 3900 },
            ],
            55,
            73,
            grades,
        )!
        expect(isOnTrackForGrade(proj, "B", grades)).toBe(true)
        expect(isOnTrackForGrade(proj, "S", grades)).toBe(false)
    })

    test("false for an unknown target grade", () => {
        const proj = projectFinalRank([{ turn: 60, rating: 5000 }], 60, 73, grades)!
        expect(isOnTrackForGrade(proj, "ZZ", grades)).toBe(false)
    })
})

File: android/app/src/test/java/com/steve1316/uma_android_automation/bot/RankProjectionTest.kt
package com.steve1316.uma_android_automation.bot

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

/** Unit tests for the pure final-rank projection math. Grade labels come from the scoring module's own (separately tested) table. */
@DisplayName("Final-rank projection")
class RankProjectionTest {
    @Test
    @DisplayName("Least-squares slope of a steady climb")
    fun testSlope() {
        assertEquals(100.0, RankProjection.slope(listOf(10 to 2000, 20 to 3000, 30 to 4000)), 1e-9)
        assertEquals(0.0, RankProjection.slope(listOf(30 to 4000)), 1e-9, "A single point has no slope")
        assertEquals(0.0, RankProjection.slope(emptyList()), 1e-9)
    }

    @Test
    @DisplayName("Projects the final score by extrapolating the fitted slope to the final turn")
    fun testProjectExtrapolation() {
        val proj = RankProjection.projectFrom(listOf(10 to 2000, 20 to 3000, 30 to 4000), totalTurns = 73)!!
        assertEquals(30, proj.currentTurn)
        assertEquals(4000, proj.currentScore)
        // slope 100 * (73-30) turns left = +4300 -> 8300.
        assertEquals(8300, proj.projectedScore)
        assertTrue(proj.currentGrade.isNotEmpty() && proj.projectedGrade.isNotEmpty(), "Grades are resolved from the score")
    }

    @Test
    @DisplayName("Never projects below the current score (rating only climbs)")
    fun testFloorAtCurrent() {
        // A dip in the last observation must not drag the projection under what's banked.
        val proj = RankProjection.projectFrom(listOf(40 to 5000, 50 to 5200, 60 to 5100), totalTurns = 73)!!
        assertTrue(proj.projectedScore >= proj.currentScore, "Projected >= current")
    }

    @Test
    @DisplayName("A single observation yields a flat projection")
    fun testSinglePoint() {
        val proj = RankProjection.projectFrom(listOf(60 to 5000), totalTurns = 73)!!
        assertEquals(0.0, proj.scorePerTurn, 1e-9)
        assertEquals(5000, proj.projectedScore)
    }

    @Test
    @DisplayName("The latest turn's observation is treated as current")
    fun testCurrentIsLatest() {
        val proj = RankProjection.projectFrom(listOf(50 to 4000, 60 to 5000, 55 to 4500), totalTurns = 73)!!
        assertEquals(60, proj.currentTurn)
        assertEquals(5000, proj.currentScore)
    }

    @Test
    @DisplayName("Empty observations project to null")
    fun testEmpty() {
        assertNull(RankProjection.projectFrom(emptyList(), totalTurns = 73))
    }

    @Test
    @DisplayName("record de-duplicates by turn and reset clears state")
    fun testRecordAndReset() {
        RankProjection.reset()
        RankProjection.record(10, 2000)
        RankProjection.record(20, 3000)
        RankProjection.record(20, 3200) // same turn re-read overwrites
        val proj = RankProjection.project(totalTurns = 73)
        assertNotNull(proj)
        assertEquals(3200, proj!!.currentScore)
        RankProjection.reset()
        assertNull(RankProjection.project(totalTurns = 73))
    }
}
